js/flexbuffers.js

```diff
@@ -0,0 +1,1015 @@
+      return;
+    }
+    const utf8 = toUTF8Array(str);
+    const length = utf8.length;
+    const newOffset = computeOffset(length + 1);
+    new Uint8Array(buffer).set(utf8, offset);
+    const stackValue = offsetStackValue(offset, flexbuffers.ValueType.KEY, flexbuffers.BitWidth.WIDTH8);
+    stack.push(stackValue);
+    keyCache[str] = stackValue;
+    offset = newOffset;
+  }
+
+  function writeStackValue(value, byteWidth) {
+    const newOffset = computeOffset(byteWidth);
+    if (value.isOffset) {
+      const relativeOffset = offset - value.offset;
+      if (byteWidth === 8 || BigInt(relativeOffset) < (1n << BigInt(byteWidth * 8))) {
+        writeInt(relativeOffset, byteWidth);
+      } else {
+        throw `Unexpected size ${byteWidth}. This might be a bug. Please create an issue https://github.com/google/flatbuffers/issues/new`
+      }
+    } else {
+      value.writeToBuffer(byteWidth);
+    }
+    offset = newOffset;
+  }
+
+  function integrityCheckOnValueAddition() {
+    if (finished) {
+      throw "Adding values after finish is prohibited";
+    }
+    if (stackPointers.length !== 0 && stackPointers[stackPointers.length - 1].isVector === false) {
+      if (stack[stack.length - 1].type !== flexbuffers.ValueType.KEY) {
+        throw "Adding value to a map before adding a key is prohibited";
+      }
+    }
+  }
+
+  function integrityCheckOnKeyAddition() {
+    if (finished) {
+      throw "Adding values after finish is prohibited";
+    }
+    if (stackPointers.length === 0 || stackPointers[stackPointers.length - 1].isVector) {
+      throw "Adding key before starting a map is prohibited";
+    }
+  }
+
+  function startVector() {
+    stackPointers.push({stackPosition: stack.length, isVector: true});
+  }
+
+  function startMap(presorted = false) {
+    stackPointers.push({stackPosition: stack.length, isVector: false, presorted: presorted});
+  }
+
+  function endVector(stackPointer) {
+    const vecLength = stack.length - stackPointer.stackPosition;
+    const vec = createVector(stackPointer.stackPosition, vecLength, 1);
+    stack.splice(stackPointer.stackPosition, vecLength);
+    stack.push(vec);
+  }
+
+  function endMap(stackPointer) {
+    if (!stackPointer.presorted) {
+      sort(stackPointer);
+    }
+    let keyVectorHash = "";
+    for (let i = stackPointer.stackPosition; i < stack.length; i += 2) {
+      keyVectorHash += `,${stack[i].offset}`;
+    }
+    const vecLength = (stack.length - stackPointer.stackPosition) >> 1;
+    if (!keyVectorCache.hasOwnProperty(keyVectorHash)) {
+      keyVectorCache[keyVectorHash] = createVector(stackPointer.stackPosition, vecLength, 2);
+    }
+    const keysStackValue = keyVectorCache[keyVectorHash];
+    const valuesStackValue = createVector(stackPointer.stackPosition + 1, vecLength, 2, keysStackValue);
+    stack.splice(stackPointer.stackPosition, vecLength << 1);
+    stack.push(valuesStackValue);
+  }
+
+  function sort(stackPointer) {
```

Comment:
Arguably, I don't fully understand the code here, but why does this roll custom sorting implementation? Is `.sort(callback)` not sufficient?

Reply:
and its bubblesort?

Reply:
It's selection sort :).
I use it based on [Wikipedia article](https://en.wikipedia.org/wiki/Selection_sort):

> Finally, selection sort is greatly outperformed on larger arrays by Θ(n log n) divide-and-conquer algorithms such as mergesort. However, insertion sort or selection sort are both typically faster for small arrays (i.e. fewer than 10–20 elements). A useful optimization in practice for the recursive algorithms is to switch to insertion sort or selection sort for "small enough" sublists.

Maps / objects often have less than 20 keys. But I am ok with switching to an in-place quick sort.

@RReverser IMHO the `.sort(callback)` is insufficient as the array looks as following:

```
[k1, v1, k2, v2, k3, v3, ....]
```

And we need to sort it based the value of `k` replacing `k` and `v` accordingly. So _logically_ it is:

```
[[k1, v1], [k2, v2], [k3, v3], ...]
```

But it is not how they are stored. Hence the custom sorting.

Reply:
I see. Perhaps it's worth changing layout of the `stack` to consist of objects (pairs) in the first place? Not sure how much of performance impact that will have, but it would have extra benefit of making it clear that those are pairs.

Reply:
Sadly it is a bit more complicated. The stack is used not only for building maps, but also for vectors and indirect values. So in case of maps we use pairs, but in case of vectors etc... it's just single StackValues.

Reply:
@aardappel should I implement in-place quick sort?

Reply:
I also implemented quick sort which is chosen when number of keys is greater then 20